### PR TITLE
Improve code coverage on collections with stream-based flows

### LINKED ISSUE #32145: Improve code coverage on collections with Stream-based flows
As part of https://github.com/dotnet/runtime/pull/2259, we deferred adding tests for every collection type due in part to keep the PR smaller (already too large to effectively review).

Specifically every collection type needs to be tests for running out of data on reads (reader.Read() returns false) and for hitting the "flush size" during write.

Whitespace can be added the the buffer to ensure each branch is hit (verified through code coverage).

For reference tests, see System.Text.Json.Serialization.Tests.StreamTests.FlushThresholdTest(), ReadSimpleObjectAsync(), and ReadAsync(). The implementation should create a custom Stream that can detect when ReadAsync()\FlushAsync() are called.

### LINKED ISSUE #34564: JsonSerializer intermittently writes invalid JSON for dictionaries in .NET 5 preview 2
I've updated a number of internal applications running .NET Core 3.1.3 to .NET 5 preview 2 (SDK version `5.0.100-preview.2.20176.6`) and have a number of previously passing tests where JSON is returned from an ASP.NET Core API using System.Text.Json that are now failing to have their contents successfully deserialized by either System.Text.Json or Newtonsoft.Json.

Exceptions similar to the following are thrown by code trying to read the JSON returned by the API:

```
System.Text.Json.JsonReaderException : '}' is an invalid start of a value. LineNumber: 503 | BytePositionInLine: 6.
   at System.Text.Json.ThrowHelper.ThrowJsonReaderException(Utf8JsonReader& json, ExceptionResource resource, Byte nextByte, ReadOnlySpan`1 bytes)
   at System.Text.Json.Utf8JsonReader.ConsumeValue(Byte marker)
   at System.Text.Json.Utf8JsonReader.ReadSingleSegment()
   at System.Text.Json.Utf8JsonReader.Read()
   at System.Text.Json.JsonDocument.Parse(ReadOnlySpan`1 utf8JsonSpan, JsonReaderOptions readerOptions, MetadataDb& database, StackRowStack& stack)
   at System.Text.Json.JsonDocument.Parse(ReadOnlyMemory`1 utf8Json, JsonReaderOptions readerOptions, Byte[] extraRentedBytes)
   at System.Text.Json.JsonDocument.Parse(Stream utf8Json, JsonDocumentOptions options)
```

```
System.Text.Json.JsonException : ':' is invalid after a value. Expected either ',', '}', or ']'. Path: $._pagination.links.first | LineNumber: 499 | BytePositionInLine: 19.
---- System.Text.Json.JsonReaderException : ':' is invalid after a value. Expected either ',', '}', or ']'. LineNumber: 499 | BytePositionInLine: 19.
   at System.Text.Json.ThrowHelper.ReThrowWithPath(ReadStack& state, JsonReaderException ex)
   at System.Text.Json.JsonSerializer.ReadCore(JsonSerializerOptions options, Utf8JsonReader& reader, ReadStack& state)
   at System.Text.Json.JsonSerializer.ReadCore(JsonReaderState& readerState, Boolean isFinalBlock, ReadOnlySpan`1 buffer, JsonSerializerOptions options, ReadStack& state)
   at System.Text.Json.JsonSerializer.ReadAsync[TValue](Stream utf8Json, Type returnType, JsonSerializerOptions options, CancellationToken cancellationToken)
--- End of stack trace from previous location ---
----- Inner Stack Trace -----
   at System.Text.Json.ThrowHelper.ThrowJsonReaderException(Utf8JsonReader& json, ExceptionResource resource, Byte nextByte, ReadOnlySpan`1 bytes)
   at System.Text.Json.Utf8JsonReader.ConsumeNextToken(Byte marker)
   at System.Text.Json.Utf8JsonReader.ConsumeNextTokenOrRollback(Byte marker)
   at System.Text.Json.Utf8JsonReader.ReadSingleSegment()
   at System.Text.Json.Utf8JsonReader.Read()
   at System.Text.Json.Serialization.Converters.ObjectDefaultConverter`1.OnTryRead(Utf8JsonReader& reader, Type typeToConvert, JsonSerializerOptions options, ReadStack& state, T& value)
   at System.Text.Json.Serialization.JsonConverter`1.TryRead(Utf8JsonReader& reader, Type typeToConvert, JsonSerializerOptions options, ReadStack& state, T& value)
   at System.Text.Json.Serialization.Converters.DictionaryDefaultConverter`2.OnTryRead(Utf8JsonReader& reader, Type typeToConvert, JsonSerializerOptions options, ReadStack& state, TCollection& value)
   at System.Text.Json.Serialization.JsonConverter`1.TryRead(Utf8JsonReader& reader, Type typeToConvert, JsonSerializerOptions options, ReadStack& state, T& value)
   at System.Text.Json.JsonPropertyInfo`1.ReadJsonAndSetMember(Object obj, ReadStack& state, Utf8JsonReader& reader)
   at System.Text.Json.Serialization.Converters.ObjectDefaultConverter`1.OnTryRead(Utf8JsonReader& reader, Type typeToConvert, JsonSerializerOptions options, ReadStack& state, T& value)
   at System.Text.Json.Serialization.JsonConverter`1.TryRead(Utf8JsonReader& reader, Type typeToConvert, JsonSerializerOptions options, ReadStack& state, T& value)
   at System.Text.Json.JsonPropertyInfo`1.ReadJsonAndSetMember(Object obj, ReadStack& state, Utf8JsonReader& reader)
   at System.Text.Json.Serialization.Converters.ObjectDefaultConverter`1.OnTryRead(Utf8JsonReader& reader, Type typeToConvert, JsonSerializerOptions options, ReadStack& state, T& value)
   at System.Text.Json.Serialization.JsonConverter`1.TryRead(Utf8JsonReader& reader, Type typeToConvert, JsonSerializerOptions options, ReadStack& state, T& value)
   at System.Text.Json.Serialization.JsonConverter`1.TryReadAsObject(Utf8JsonReader& reader, Type typeToConvert, JsonSerializerOptions options, ReadStack& state, Object& value)
   at System.Text.Json.JsonSerializer.ReadCore(JsonSerializerOptions options, Utf8JsonReader& reader, ReadStack& state)
```

I've yet to be able to create a reliable isolated repro (but I will try further to do so), but there appear to be some common themes to the failing tests in these two different applications, though some may be coincidences:
  1. The written JSON is malformed inside an object which is the value of a dictionary.
  1. A single failing test run by itself passes, suggesting that something goes wrong over time rather than always being broken.
  1. Tests that fail appear to be one or more of a set of xunit theories running mostly the same code with slightly different inputs or expected outputs, suggesting some sort of shared state being corrupted.
  1. Which tests fail varies between test runs.
  1. The invalid JSON is approximately 450-520 lines into the returned JSON response.

Below is a simplified invalid JSON string returned by the API showing the malformed part of the response within the wider context of the overall response:

```json
{
  "consumerId": "456",
  "tenant": "uk",
  "orders": [
    {
      "id": "abc"
    },
    {
      "id": "def"
    }
  ],
  "_links": {
    "self": {
      "href": "",
      "templated": false
    }
  },
  "_pagination": {
    "total": null,
    "count": 2,
    "token": "",
    "links": {
      "self": {
        "href": "",
        "templated": false
      },
      "first": {
        "href": "",
        "templated": false
      },
      "prev": {
        "href": "",
        "templated": false,
        "prev": 
      },
      "next": {
        "href": "",
        "templated": false
      }
    }
  },
  "timestamp": "2017-06-05T17:20:00\u002B00:00"
}
```

The invalid part is here, where a dangling `"prev": ` is written before the end of the object before the penultimate object in the dictionary is written:

```json
      "prev": {
        "href": "",
        "templated": false,
        "prev": 
      }
```

The relevant parts of the serialized object model for this part of the JSON are:

```csharp
public class Link
{
    [JsonPropertyName("href")]
    public string Url { get; set; }

    [JsonPropertyName("templated")]
    public bool IsTemplated { get; set; }
}

public class PaginationInfo
{
    public PaginationInfo()
    {
        Links = new Dictionary<string, Link>(StringComparer.Ordinal);
    }

    [JsonPropertyName("total")]
    public long? Total { get; set; }

    [JsonPropertyName("count")]
    public long Count { get; set; }

    [JsonPropertyName("token")]
    public string Token { get; set; }

    [JsonPropertyName("links")]
    public IDictionary<string, Link> Links { get; set; }
}

public class Order
{
    [JsonPropertyName("id")]
    public string Id { get; set; }
}

public class OrderHistory
{
    public OrderHistory()
    {
        Links = new Dictionary<string, Link>(StringComparer.Ordinal);
        Orders = Array.Empty<Order>();
    }

    [JsonPropertyName("consumerId")]
    public string ConsumerId { get; set; }

    [JsonPropertyName("tenant")]
    public string Tenant { get; set; }

    [JsonPropertyName("orders")]
    public ICollection<Order> Orders { get; set; }

    [JsonPropertyName("_links")]
    public IDictionary<string, Link> Links { get; set; }

    [JsonPropertyName("_pagination")]
    public PaginationInfo Pagination { get; set; }

    [JsonPropertyName("timestamp")]
    public DateTimeOffset Timestamp { get; set; }
}
```

Looking at GitHub's suggested similar issues and looking at the source history, it's possible this issue is related to #32523.

### LINKED ISSUE #35278: Issue when combining ReferenceHandling.Preserve and DeserializeAsync
Originally reported on https://github.com/dotnet/runtime/issues/30820#issuecomment-613272109.

Code to reproduce the issue:
```cs
public static async Task ReadMoreDataOnMetadataValueAsync()
{
    // 1..7
    string json = "{\"$id\":";

    // 8..17 - Problem happpens when we step in this portion of the JSON payload.
    json += "\"90123456\"";

    // 18..32
    json += ",\"foo\":\"value\"}";

    var stream = new MemoryStream(Encoding.UTF8.GetBytes(json));

    var opts = new JsonSerializerOptions
    {
        DefaultBufferSize = 16,
        ReferenceHandling = ReferenceHandling.Preserve
    };

    var foo = await JsonSerializer.DeserializeAsync<Foo>(stream, opts);
}
```

Error:
```
System.Text.Json.JsonException : The '$id' and '$ref' metadata properties must be JSON strings. Current token type is 'PropertyName'. Path: $.$id | LineNumber: 0 | BytePositionInLine: 24.
        Stack Trace:
          C:\repos\runtime\src\libraries\System.Text.Json\src\System\Text\Json\ThrowHelper.Serialization.cs(92,0): at System.Text.Json.ThrowHelper.ThrowJsonException(String message)
          C:\repos\runtime\src\libraries\System.Text.Json\src\System\Text\Json\ThrowHelper.Serialization.cs(449,0): at System.Text.Json.ThrowHelper.ThrowJsonException_MetadataValueWasNotString(JsonTokenType tokenType)
          C:\repos\runtime\src\libraries\System.Text.Json\src\System\Text\Json\Serialization\JsonSerializer.Read.HandleMetadata.cs(124,0): at System.Text.Json.JsonSerializer.ResolveMetadata(JsonConverter converter, Utf8JsonReader& reader, ReadStack& state)
          C:\repos\runtime\src\libraries\System.Text.Json\src\System\Text\Json\Serialization\Converters\Object\ObjectDefaultConverter.cs(81,0): at System.Text.Json.Serialization.Converters.ObjectDefaultConverter`1.OnTryRead(Utf8JsonReader& reader, Type typeToConvert, JsonSerializerOptions options, ReadStack& state, T& value)
          C:\repos\runtime\src\libraries\System.Text.Json\src\System\Text\Json\Serialization\JsonConverterOfT.cs(182,0): at System.Text.Json.Serialization.JsonConverter`1.TryRead(Utf8JsonReader& reader, Type typeToConvert, JsonSerializerOptions options, ReadStack& state, T& value)
          C:\repos\runtime\src\libraries\System.Text.Json\src\System\Text\Json\Serialization\JsonConverterOfT.ReadCore.cs(61,0): at System.Text.Json.Serialization.JsonConverter`1.ReadCore(Utf8JsonReader& reader, JsonSerializerOptions options, ReadStack& state)
          C:\repos\runtime\src\libraries\System.Text.Json\src\System\Text\Json\Serialization\JsonSerializer.Read.Helpers.cs(26,0): at System.Text.Json.JsonSerializer.ReadCore[TValue](JsonConverter jsonConverter, Utf8JsonReader& reader, JsonSerializerOptions options, ReadStack& state)
          C:\repos\runtime\src\libraries\System.Text.Json\src\System\Text\Json\Serialization\JsonSerializer.Read.Stream.cs(236,0): at System.Text.Json.JsonSerializer.ReadCore[TValue](JsonReaderState& readerState, Boolean isFinalBlock, ReadOnlySpan`1 buffer, JsonSerializerOptions options, ReadStack& state, JsonConverter converterBase)
          C:\repos\runtime\src\libraries\System.Text.Json\src\System\Text\Json\Serialization\JsonSerializer.Read.Stream.cs(168,0): at System.Text.Json.JsonSerializer.ReadAsync[TValue](Stream utf8Json, Type returnType, JsonSerializerOptions options, CancellationToken cancellationToken)
          C:\repos\runtime\src\libraries\System.Text.Json\tests\Serialization\ReferenceHandlingTests.cs(60,0): at System.Text.Json.Serialization.Tests.ReferenceHandlingTests.ReadMoreDataOnMetadataValueAsync()
```

The reason why this happens is that when we step into the token `"90123456"` we exceed the deserializer's buffer size (16), so we have to unwind the call stack in order to read more data from the stream. When we step back into the code that should read the string token (`"90123456"`), we call `reader.Read` again and hence the reader moves to the next token which is a property (`"foo"`). That should not happen when stepping back where we left since we didn't complete processing the previous token in the first place. 

One possible solution could be to avoid calling `reader.Read` again when we step back into the `HendleMetadata` method after more data is loaded into the deserializer buffer.

### PR DESCRIPTION
**Pull request by layomia**
- Fixes https://github.com/dotnet/runtime/issues/32145 (Improve code coverage on collections with Stream-based flows).

    Collection converters are fully covered (line & branch). There are still some gaps in `IEnumerableDefaultConverter`, `ObjectDefaultConverter`, and `DictionaryDefaultConverter`. I plan to address those in a follow up as apart of https://github.com/dotnet/runtime/issues/32341.

    **Before**
before

    **After**
after

- Fixes https://github.com/dotnet/runtime/issues/34564 (JsonSerializer intermittently writes invalid JSON for dictionaries in .NET 5 preview 2).

    Some of the dictionary converters didn't have logic need to resume when the reader fails to read the entire key.

- Fixes https://github.com/dotnet/runtime/issues/35278 (Issue when combining ReferenceHandling.Preserve and DeserializeAsync).

    There were bugs with parsing metadata related to reference handling on deserialization: "id", "ref", "values". The responsible method did not have logic required to resume when `reader.Read()` fails.

- Adds misc branch coverage tests for deserializing collections.

### CODE REVIEW

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Collection/IDictionaryConverter.cs [hunk 1]
```diff
@@ -68,31 +65,41 @@ protected internal override bool OnWriteResume(Utf8JsonWriter writer, TCollectio
...
             {
-                Debug.Assert(state.Current.CollectionEnumerator is IDictionaryEnumerator);
                 enumerator = (IDictionaryEnumerator)state.Current.CollectionEnumerator;
             }
 
             JsonConverter<object?> converter = GetValueConverter(ref state);
             do
             {
-                if (enumerator.Key is string key)
+                if (ShouldFlush(writer, ref state))
```
Comment (steveharter) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Collection/IDictionaryConverter.cs:74]: So this was the only collection we didn't apply the "ShouldFlush" pattern to?

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.HandleMetadata.cs [hunk 2]
```diff
@@ -94,13 +99,27 @@ public static partial class JsonSerializer
                 }
             }
 
-            if (state.Current.ObjectState == StackFrameObjectState.MetadataRefProperty)
+            if (state.Current.ObjectState == StackFrameObjectState.MetadataTryReadRefValue)
```
Comment (steveharter) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.HandleMetadata.cs:102]: Is it necessary to have two states here (one would suffice?). I see in other logic below they may have different paths, but it's hard to determine why.

At least we could just use an OR here instead of two branches.

File: src/libraries/System.Text.Json/tests/Serialization/Stream.Collections.cs [hunk 3]
```diff
@@ -0,0 +1,341 @@
...
+
+                    var optionsWithPreservedReferenceHandling = new JsonSerializerOptions(options)
+                    {
+                        ReferenceHandling = ReferenceHandling.Preserve
+                    };
+
+                    Type elementType = typeof(TElement);
+
+                    Task[] tasks = new Task[2];
+                    tasks[0] = Task.Run(async () => await PerformSerialization(obj, type, elementType, options));
```
Comment (steveharter) [inline src/libraries/System.Text.Json/tests/Serialization/Stream.Collections.cs:51]: So we are also testing multiple threads here (in addition to faulting on the buffer) otherwise we'd have two `await`s instead of `WaitAll`. Did you encounter issues with shared state anywhere from multiple threads?

File: src/libraries/System.Text.Json/tests/Serialization/Stream.Collections.cs [hunk 4]
```diff
@@ -0,0 +1,341 @@
...
+            {
+                return Activator.CreateInstance(type, new object[] { GetArr_TypedElements(value) });
+            }
+            else
+            {
+                return Activator.CreateInstance(type, new object[] { GetArr_BoxedElements(value) });
+            }
+        }
+
+        private static void AssertJsonEqual(string expected, string actual)
```
Comment (steveharter) [inline src/libraries/System.Text.Json/tests/Serialization/Stream.Collections.cs:144]: Is this necessary due to reflection ordering being nondeterministic?

I wonder if this should be moved to a shared test utility class.

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Collection/IDictionaryConverter.cs (hunk 1 above)
Comment (layomia) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Collection/IDictionaryConverter.cs:74]: Yup.

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.HandleMetadata.cs (hunk 2 above)
Comment (layomia) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.HandleMetadata.cs:102]: As we discussed offline with @Jozkee, we do need two states. Ultimately, finding "ref" means we can just set `state.Current.ReturnValue` to a previously seen reference (plus some validation) and exit, while finding "id" would mean we should continue reading JSON properties/read array "Values" metadata...

File: src/libraries/System.Text.Json/tests/Serialization/Stream.Collections.cs (hunk 3 above)
Comment (layomia) [reply src/libraries/System.Text.Json/tests/Serialization/Stream.Collections.cs:51]: [As discussed offline] the tasks use separate options instances, so only faulting on the buffer is tested here. I've changed this to use two `await`s.

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Collection/ImmutableDictionaryOfStringTValueConverter.cs [hunk 5]
```diff
@@ -44,8 +41,7 @@ protected internal override bool OnWriteResume(Utf8JsonWriter writer, TCollectio
             }
             else
             {
-                Debug.Assert(state.Current.CollectionEnumerator is Dictionary<string, TValue>.Enumerator);
-                enumerator = (Dictionary<string, TValue>.Enumerator)state.Current.CollectionEnumerator;
+                enumerator = (IEnumerator<KeyValuePair<string, TValue>>)state.Current.CollectionEnumerator;
```
Comment (Jozkee) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Collection/ImmutableDictionaryOfStringTValueConverter.cs:44]: What's the reason for this change?

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Object/ObjectWithParameterizedConstructorConverter.cs [hunk 6]
```diff
@@ -176,7 +176,7 @@ private void ReadConstructorArguments(ref ReadStack state, ref Utf8JsonReader re
 
                     if (!(jsonParameterInfo!.ShouldDeserialize))
                     {
-                        reader.TrySkip();
+                        reader.Skip();
```
Comment (Jozkee) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Object/ObjectWithParameterizedConstructorConverter.cs:179]: Before this we were silently failing (i.e. not throwing) when there was invalid JSON as the value of a parameter that wasn't a candidate for deserialization, right? Did you add a test for this?

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.HandleMetadata.cs [hunk 7]
```diff
@@ -18,14 +18,19 @@ public static partial class JsonSerializer
             ref Utf8JsonReader reader,
             ref ReadStack state)
         {
-            if (state.Current.ObjectState < StackFrameObjectState.MetadataPropertyName)
+            if (state.Current.ObjectState < StackFrameObjectState.MetadataTryReadName)
             {
+                state.Current.ObjectState = StackFrameObjectState.MetadataReadName;
```
Comment (Jozkee) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.HandleMetadata.cs:23]: Maybe we should rename `MetadataTryReadName` to `MetadataTryReadNameOrEndObject`.

Comment (Jozkee) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.HandleMetadata.cs:23]: Can we add a helper method for all of these "try read and set next state"?

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.HandleMetadata.cs [hunk 8]
```diff
@@ -18,14 +18,19 @@ public static partial class JsonSerializer
             ref Utf8JsonReader reader,
```
Comment (Jozkee) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.HandleMetadata.cs:18]: How's the test coverage for `ResolveMetadata`?

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Object/ObjectWithParameterizedConstructorConverter.cs (hunk 6 above)
Comment (layomia) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Object/ObjectWithParameterizedConstructorConverter.cs:179]: No -  this code path applies only when `reader.IsFinalBlock` is true, so `reader.TrySkip` was always `true` here. This code path has full test coverage.

It would be good to have tests focused on encountering invalid JSON when deserializing objects (with both parameterless & parameterized ctors) & when deserializing when `options.ReferenceHandling.Preserve` is active, similar to work done in https://github.com/dotnet/corefx/pull/42254 & https://github.com/dotnet/corefx/pull/42066. I logged an issue to track this - https://github.com/dotnet/runtime/issues/35517.

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.HandleMetadata.cs (hunk 8 above)
Comment (layomia) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.HandleMetadata.cs:18]: It does have some gaps - Line 84.35%, Branch 92.64%. Addressing this is out of scope for this PR - we should address this as part of https://github.com/dotnet/runtime/issues/32341 & likely https://github.com/dotnet/runtime/issues/35517.

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Collection/ImmutableDictionaryOfStringTValueConverter.cs (hunk 5 above)
Comment (layomia) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Collection/ImmutableDictionaryOfStringTValueConverter.cs:44]: The initial code is incorrect, as the enumerator for these collections can't be cast to `Dictionary<string, TValue>.Enumerator`. We didn't have coverage for this scenario (resuming after fetching more data during async deserialization of immutable dictionaries) prior to this PR. We'd have leaked an `InvalidCastException`.

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/StackFrameObjectState.cs [hunk 9]
```diff
@@ -14,21 +14,22 @@ internal enum StackFrameObjectState : byte
 
         StartToken,
 
-        MetadataTryReadName, // Try to move the reader to the the first $id or $ref.
-        MetadataReadName, // Read the first $id or $ref.
+        ReadAheadNameOrEndObject, // Try to move the reader to the the first $id, $ref, or the EndObject token.
```
Comment (layomia) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/StackFrameObjectState.cs:17]: Removed the `Metadata` prefix on these values as the names were getting too long to read/understand quickly. I think the intent is clear in the respective call sites.

File: src/libraries/System.Text.Json/tests/Serialization/Stream.Collections.cs [hunk 10]
```diff
@@ -17,235 +18,247 @@ namespace System.Text.Json.Serialization.Tests
...
+            else
+            {
+                return Activator.CreateInstance(type);
+            }
         }
 
-        private static void AssertJsonEqual(JsonElement expected, JsonElement actual)
+        private static string GetPayloadWithWhiteSpace(string json) => json.Replace(" ", new string(' ', 4));
+
+        private const int NumElements = 15;
```
Comment (layomia) [inline src/libraries/System.Text.Json/tests/Serialization/Stream.Collections.cs:172]: Reduced the number of collection elements to test for - this reduces the time to run the entire test suite from ~45s to ~28s, with no changes to code coverage.

File: src/libraries/System.Text.Json/tests/Serialization/Stream.Collections.cs [hunk 11]
```diff
@@ -17,235 +18,247 @@ namespace System.Text.Json.Serialization.Tests
     public static partial class StreamTests
```
Comment (layomia) [inline src/libraries/System.Text.Json/tests/Serialization/Stream.Collections.cs:18]: Aside from applying review feedback, changes here are mostly cosmetic to make the code a bit cleaner & easier to read.

I also remove asserts on serialized dictionaries with `ReferenceHandling.Preserve` since reflection ordering is not guaranteed -- `ref`/`id` metadata values might not match up.

File: src/libraries/System.Text.Json/tests/JsonTestHelper.cs [hunk 12]
```diff
@@ -814,5 +814,57 @@ public static string NormalizeLineEndings(this string value)
...
+            using JsonDocument actualDom = JsonDocument.Parse(actual);
+            AssertJsonEqual(expectedDom.RootElement, actualDom.RootElement);
+        }
+
+        private static void AssertJsonEqual(JsonElement expected, JsonElement actual)
+        {
+            JsonValueKind valueKind = expected.ValueKind;
+            Assert.Equal(valueKind, actual.ValueKind);
+
+            switch (valueKind)
```
Comment (layomia) [inline src/libraries/System.Text.Json/tests/JsonTestHelper.cs:830]: As more tests use these helper methods, we'll need to add more cases here; e.g. `case JsonValueKind.Number`.

**Review by layomia — COMMENTED** [review]
Adding some notes about the new commit https://github.com/dotnet/runtime/pull/35046/commits/14ab70a9e1d09f06a38e32009212f36e143072ab.

File: src/libraries/System.Text.Json/tests/Serialization/Stream.Collections.cs (hunk 4 above)
Comment (layomia) [reply src/libraries/System.Text.Json/tests/Serialization/Stream.Collections.cs]: Yes, due to nondeterministic reflection.

I've added this to `JsonTestHelper` as this is useful when verifying serialized payloads with multiple properties (nicer than bunch of `Assert.Contains` statements). Also see https://github.com/dotnet/runtime/pull/35046#discussion_r416060333.

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Object/ObjectWithParameterizedConstructorConverter.cs (hunk 6 above)
Comment (Jozkee) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Object/ObjectWithParameterizedConstructorConverter.cs:179]: > This code path has full test coverage.

What's the reason for this change, then?

File: src/libraries/System.Text.Json/tests/Serialization/Stream.Collections.cs (hunk 11 above)
Comment (Jozkee) [reply src/libraries/System.Text.Json/tests/Serialization/Stream.Collections.cs:18]: > reflection ordering is not guaranteed -- ref/id metadata values might not match up.

Hmm, what does reflection ordering has to do with dictionary KeyValuePairs and metadata? 
Doesn't reflection ordering only affects the order of object's properties?

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Object/ObjectWithParameterizedConstructorConverter.cs (hunk 6 above)
Comment (layomia) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Object/ObjectWithParameterizedConstructorConverter.cs:179]: Are you asking why was changed in this PR? I noticed it while skimming through the files for places where we throw when `reader.TokenType != JsonTokenType.PropertyName` (in some cases, `Debug.Assert` is fine). I thought to change it since I have context here.

---
Otherwise,

`TrySkip` is meant for when the reader might not have the data needed for the children to be skipped successfully. We know that's not the case here.

Using `Skip` provides extra validation for this assumption because `InvalidOperationException` is thrown if `IsFinalBlock` is false.

Comment (Jozkee) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Object/ObjectWithParameterizedConstructorConverter.cs:179]: > Using Skip provides extra validation for this assumption because InvalidOperationException is thrown if IsFinalBlock is false.

This seems reasonable to me, and its the answer to what I was trying to ask. Thanks.

File: src/libraries/System.Text.Json/tests/Serialization/Stream.Collections.cs (hunk 11 above)
Comment (layomia) [reply src/libraries/System.Text.Json/tests/Serialization/Stream.Collections.cs:18]: I meant the order of elements returned by `.GetEnumerator()` for dictionaries, which is the order we serialize & check for metadata. The order is not guaranteed IIRC. I hit a few test failures because of this locally.

Comment (Jozkee) [reply src/libraries/System.Text.Json/tests/Serialization/Stream.Collections.cs:18]: > which is the order we serialize & check for metadata.

We don't check for metadata on serialization, we append metadata to the resulting JSON objects and collections; metadata properties are never part of the CLR object.

Comment (Jozkee) [reply src/libraries/System.Text.Json/tests/Serialization/Stream.Collections.cs:18]: Could it be that the issue that you are facing is caused by a race condition bug in the test?

Comment (layomia) [reply src/libraries/System.Text.Json/tests/Serialization/Stream.Collections.cs:18]: > > which is the order we serialize & check for metadata.
> 
> We don't check for metadata on serialization

check for **references***

We check if we have seen the instance before in order to assign it metadata, and that is done in the order returned by the enumerator.

The dictionary element that takes an "id" of "1" when we serialize may not be the same one that takes it during the next serialization.

> Could it be that the issue that you are facing is caused by a race condition bug in the test?

I don't believe so.

Comment (Jozkee) [reply src/libraries/System.Text.Json/tests/Serialization/Stream.Collections.cs:18]: As we discussed offline, a Dictionary enumeration returns elements in a non-deterministic manner on Serialization time, which will cause that its elements don't get the same id on each call. So condition on line 99 `if(!(DictionaryTypes<TElement>().Contains(type) && options.ReferenceHandling == ReferenceHandling.Preserve))` is correct.

Comment (layomia) [reply src/libraries/System.Text.Json/tests/Serialization/Stream.Collections.cs:18]: After this PR is merged, I'll open an issue discussing ways we can run the test for the `if(DictionaryTypes<TElement>().Contains(type) && options.ReferenceHandling == ReferenceHandling.Preserve)` case. One way is to override the GetEnumerator method and force a particular order.

**Review by steveharter — APPROVED** [review]
